### PR TITLE
Added support to other naming strategy

### PR DESCRIPTION
**Pull request by eliax1996**
Before that commit the other strategies where existing but not effectively usable/tested in the codebase.
This pr adds the support and set tests to effectively test the naming strategies.
During the development I've also added a little bit more typing over the files I've touched to improve the codebase

### CODE REVIEW

File: karapace/config.py [hunk 1]
```diff
@@ -269,12 +281,14 @@ def validate_config(config: Config) -> None:
             f"Invalid master election strategy: {master_election_strategy}, valid values are {valid_strategies}"
         ) from None
 
-    name_strategy = config["name_strategy"]
+    deafault_name_strategy = config["name_strategy"]
```
Comment (tvainika) [inline karapace/config.py:284]: typo: deafault -> default

File: karapace/serialization.py [hunk 2]
```diff
@@ -225,6 +227,25 @@ async def close(self):
         await self.client.close()
 
 
+def get_subject_name(
```
Comment (tvainika) [inline karapace/serialization.py:230]: I see that you only moved here code around, not changing. However now looking at this, is it correct code at all? For me it looks it takes only namespace of record, but is missing record name itself? `naming_strategy` functions take `record_name` as argument, not `namespace`. This is also reflected in the test you added. Are those expected values correct in that test is relevant question in this context.

File: tests/unit/test_serialization.py [hunk 3]
```diff
@@ -310,3 +314,21 @@ async def test_deserialization_fails(default_config_path):
         await deserializer.deserialize(enc_bytes)
 
     assert mock_registry_client.method_calls == [call.get_schema_for_id(1)]
+
+
+@pytest.mark.parametrize(
+    "expected_subject,strategy,subject_type",
+    (
+        (Subject("foo-key"), NameStrategy.topic_name, SUBJECT_VALID_POSTFIX[0]),
```
Comment (tvainika) [inline tests/unit/test_serialization.py:322]: See comment related to this test.

But also this test is more readable if there is no import of `SUBJECT_VALID_POSTFIX` but instead each parametrized test here refers either to `SubjectType.key` or `SubjectType.value`, IMHO.

File: karapace/serialization.py [hunk 4]
```diff
@@ -243,37 +264,17 @@ def __init__(
         else:
             registry_url = f"http://{self.config['registry_host']}:{self.config['registry_port']}"
             registry_client = SchemaRegistryClient(registry_url, session_auth=session_auth)
-        name_strategy = config.get("name_strategy", "topic_name")
-        self.subject_name_strategy = NAME_STRATEGIES.get(name_strategy, topic_name_strategy)
         self.registry_client: Optional[SchemaRegistryClient] = registry_client
         self.ids_to_schemas: Dict[int, TypedSchema] = {}
         self.ids_to_subjects: MutableMapping[int, List[Subject]] = TTLCache(maxsize=10000, ttl=600)
         self.schemas_to_ids: Dict[str, SchemaId] = {}
+        self._topic_strategy_cache: MutableMapping[TopicName, NameStrategy] = TTLCache(maxsize=10000, ttl=600)
```
Comment (tvainika) [inline karapace/serialization.py:271]: Does this belong to this PR or some other work?

**Review by tvainika — CHANGES_REQUESTED** [review]
See inline the fundamental question how record name strategy should work to be clarified

File: karapace/serialization.py (hunk 4 above)
Comment (eliax1996) [reply karapace/serialization.py:271]: ops, another piece I forgot. Thanks

File: tests/unit/test_serialization.py (hunk 3 above)
Comment (eliax1996) [reply tests/unit/test_serialization.py:322]: Yeah, you are right. Initially I've wrote the test as you are suggesting, then I thought while adding the `SUBJECT_VALID_POSTFIX` variable, shouldn't we use the same also in test? But it looks terrible, going to revert

File: karapace/serialization.py (hunk 2 above)
Comment (eliax1996) [reply karapace/serialization.py:230]: > is it correct code at all? For me it looks it takes only namespace of record, but is missing record name itself? 
good point, I've just moved stuff around and created a test to keep the behaviour. We need to find a real case where this is done and compare it to our implementation, I'm looking for it now

File: karapace/serialization.py [hunk 5]
```diff
@@ -59,37 +62,59 @@ class SchemaUpdateError(SchemaError):
...
+
 
+def record_name_strategy(
+    topic_name: str,  # pylint: disable=unused-argument,
+    record_name: str | None,
+    subject_type: SubjectType,  # pylint: disable=unused-argument
+) -> Subject:
+    if record_name is None:
+        raise InvalidRecord(
+            "The provided record do not have a valid `record_name`, use another naming strategy or fix the schema"
```
Comment (giuseppelillo) [inline karapace/serialization.py:84]:
```suggestion
            "The provided record does not have a valid `record_name`, use another naming strategy or fix the schema"
```

Comment (eliax1996) [reply karapace/serialization.py:84]: maybe directly `doesn't`?

**Review by giuseppelillo — COMMENTED** [review]
Overall LGTM. Could you also add a test on how the record and topic-record naming strategy are supposed to work with an Avro schema without namespace?

File: karapace/serialization.py (hunk 5 above)
Comment (giuseppelillo) [reply karapace/serialization.py:84]: Whatever you prefer :D

**Review by tvainika — CHANGES_REQUESTED** [review]
LGTM except documentation change is not correct...

File: README.rst [hunk 6]
```diff
@@ -460,7 +460,7 @@ Keys to take special care are the ones needed to configure Kafka and advertised_
      - ``runtime``
      - Runtime directory for the ``protoc`` protobuf schema parser and code generator
    * - ``name_strategy``
-     - ``topic_name``
+     - ``topic_name``, ``record_name``, ``topic_record_name``
```
[the remark this answers is not in the document]
Comment (eliax1996) [reply README.rst:463]: 😵‍💫 pushing both suggestion/fix asap

**Review by tvainika — APPROVED** [review]

**Review by giuseppelillo — APPROVED** [review]

**Pull request merged**